### PR TITLE
More GdbServer improvements

### PR DESCRIPTION
**Pull request by Sonicadvance1**
Nothing too crazy in here, trying to avoid some of the bigger changes.
A couple of stand-out things

- Adds some lldb specific query commands.
- Set the thread name of gdbserver. Easier to pick out if a thread hung.
- Try binding twice, unlinking the unix domain socket path in between
- Poll the listen socket instead of waiting in accept. Removing a hang that could occur.

### CODE REVIEW

File: FEXCore/include/FEXCore/Utils/StringUtils.h [hunk 1]
```diff
@@ -28,4 +28,41 @@ namespace FEXCore::StringUtils {
...
 
+  namespace StringHasher {
+    struct Hasher {
+      // Jenkins one-at-a-time function.
+      // Based off of https://en.wikipedia.org/wiki/Jenkins_hash_function
+      // citing: https://www.drdobbs.com/database/algorithm-alley/184410284
+      // and citing: https://www.burtleburtle.net/bob/hash/doobs.html
+      //
+      // Good enough constexpr hasher.
+      uint32_t constexpr operator()(const char* str) const {
```
Comment (lioncash) [inline FEXCore/include/FEXCore/Utils/StringUtils.h:39]:
```suggestion
      size_t constexpr operator()(const char* str) const noexcept {
```

would make this match the interface requirements of std::hash

File: FEXCore/include/FEXCore/Utils/StringUtils.h [hunk 2]
```diff
@@ -28,4 +28,41 @@ namespace FEXCore::StringUtils {
...
+          ++str;
+        }
+        hash += hash << 3;
+        hash ^= hash >> 11;
+        hash += hash << 15;
+        return hash;
+      };
+
+      template<typename T>
+      uint32_t constexpr operator()(const T& str) const {
```
Comment (lioncash) [inline FEXCore/include/FEXCore/Utils/StringUtils.h:54]: Ditto

File: FEXCore/include/FEXCore/Utils/StringUtils.h [hunk 3]
```diff
@@ -28,4 +28,41 @@ namespace FEXCore::StringUtils {
     return RightTrim(LeftTrim(std::move(String), TrimTokens), TrimTokens);
   }
 
+  namespace StringHasher {
```
Comment (neobrain) [inline FEXCore/include/FEXCore/Utils/StringUtils.h:31]: This doesn't seem to be related to the other changes. Did you mean to include any other patches?

If not, reviewing this along with code that actually uses it seems more sensible.

Comment (Sonicadvance1) [reply FEXCore/include/FEXCore/Utils/StringUtils.h:31]: This is used in GdbServer.cpp

Comment (neobrain) [reply FEXCore/include/FEXCore/Utils/StringUtils.h:31]: Ah, yes... "Large diffs are not rendered by default" got me once again :(

File: Source/Tools/FEXLoader/LinuxSyscalls/GdbServer.cpp [hunk 4]
```diff
@@ -661,84 +671,84 @@ GdbServer::HandledPacketType GdbServer::handleXfer(const fextl::string &packet)
...
-        // Thread id is in hex without 0x prefix
-        const auto ThreadName = getThreadName(Thread->ThreadManager.GetTID());
-        ss << "<thread id=\"" << std::hex << Thread->ThreadManager.GetTID() << "\"";
-        if (!ThreadName.empty()) {
-          ss << " name=\"" << ThreadName << "\"";
-        }
-        ss << "/>\n";
+  using namespace FEXCore::StringUtils;
+  using namespace FEXCore::StringUtils::StringHasher;
+  switch (StringHasher::Hash(object)) {
```
[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply Source/Tools/FEXLoader/LinuxSyscalls/GdbServer.cpp:676]: I can understand not wanting the idiom. I'll remove it.

File: Source/Tools/FEXLoader/LinuxSyscalls/GdbServer.cpp [hunk 5]
```diff
@@ -1312,13 +1416,21 @@ void GdbServer::OpenListenSocket() {
...
+  };
+
   // Bind the socket to the path
-  int Result = bind(ListenSocket, reinterpret_cast<struct sockaddr*>(&addr), SizeOfAddr);
-  if (Result == -1) {
-    LogMan::Msg::EFmt("[GdbServer] Couldn't bind AF_UNIX socket '{}': {} {}\n", addr.sun_path, errno, strerror(errno));
-    close(ListenSocket);
-    ListenSocket = -1;
-    return;
+  if (!TryBind(ListenSocket, &addr, SizeOfAddr)) {
```
Comment (neobrain) [inline Source/Tools/FEXLoader/LinuxSyscalls/GdbServer.cpp:1425]: Using a lambda here for just one repetition adds more noise than it really saves. Maybe a for loop does the trick more easily?

```cpp
for (int attempt = 0; attempt < 2; ++attempt) {
    ...
}
```

**Review by neobrain — APPROVED** [review]
Some minor comments, looks good otherwise 👍

File: Source/Tools/FEXLoader/LinuxSyscalls/GdbServer.cpp [hunk 6]
```diff
@@ -1312,13 +1424,21 @@ void GdbServer::OpenListenSocket() {
...
+    // This can happen periodically with execve. unlink the path and try again.
```
[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply Source/Tools/FEXLoader/LinuxSyscalls/GdbServer.cpp:1427]: Added a comment that the PID is reused on execve and that FEX likely started a gdbserver before execve happened.
